Extract websocket connection and message handling out of LAORepository (#806)
dedis/popstellar: fe2-android/docs/README.md

of [multiple layers](https://docs.google.com/document/d/1AeV7JX_SJ30mu9PIwmz24UkIi3jCo6NSYA0sPxdbscU)
and you need to be familiar with them to understand how communication happens.

The [remote module](https://github.com/dedis/popstellar/tree/master/fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote)
contains most of the network communication logic, it is organized as follows:\
The `GlobalNetworkManager` is responsible for holding, creating and destroying WebSocket connections.\
It uses the `ConnectionFactory` to produce new `Connection` objects using
the [Scarlet](https://github.com/Tinder/Scarlet) library.

> The `Connection` is passed to a `LAOConnectionManager` that is responsible for handling the JsonRPC layer of the

Reviewer: technically it handles also WebSockets, through Scarlet, no? we might want to mention that, since it shows up in the diagram.

Author: From the `LAOConnectionManager` point of view, there is nothing to do with the actual websocket it only sees a `Connection` that exposes an input message stream, input event stream and a way to send messages.

I will apply the name changes that you spotted and it should make things clearer.

> protocol. It exposes the protocol API with the `MessageSender` interface that can be retrieved by `ViewModel`though

Reviewer: ```suggestion
protocol. It exposes the protocol API with the `MessageSender` interface that can be retrieved by `ViewModel` though
```

the `GlobalNetworkManager`.

For more information on sending messages on the network, please refer to
the [remote module](https://github.com/dedis/popstellar/tree/master/fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote)
and the `ViewModel` classes. Also, make sure you have a solid understanding
of [JSON-RPC](https://www.jsonrpc.org/specification),
the [Protocol Specifications](https://docs.google.com/document/d/1fyNWSPzLhM6W9V0VTFf2waMLiJGcscy7wa4bQlLkySM)
and their actual implementation in the [protocol schemas](https://github.com/dedis/popstellar/tree/master/protocol)

### Getting messages over the wire

Once it is clear how to send messages over the wire, it is important to turn one's attention to
receiving them.

> Because receiving messages over the network and processing them to update the application state are very different steps
> involving completely unrelated logic.

Reviewer: Not your fault, but this is not a sentence, there's no verb 😅 
```suggestion
Receiving messages over the network and processing them to update the application state are very different steps. As such, they involve completely unrelated logic.
```
but is it still true? to some extent, I feel it's not the case anymore and we could remove these 2 lines altogether.

> On the networking side, the embedded Scarlet service handles the upcoming websocket messages and transform them into
> Java Object using Gson as the Json Parser.

Reviewer: I like how you've structured the coming lines as "following a message" through the network stack. I'd make it even more explicit.

```suggestion
When the network dispatches a WebSocket message to the application, the embedded Scarlet service handles it and transforms it into
a Java Object using Gson as the Json Parser.
```

> `LAONetworkManager` then hands them to the `MessageHandler` that will use the application's `DataRegistry` to pass the
> Data message to the right handler. The `handler` submodule is responsible for processing different kind of messages,

Reviewer: ```suggestion
The `LAONetworkManager` then hands the message over to the `MessageHandler`, which will in turn use the application's `DataRegistry` to pass the
Data message to the right handler. The `handler` submodule is responsible for processing different kind of messages,
```
